### PR TITLE
Changed $mappings to $entities

### PR DESCRIPTION
**Pull request by chrimc62**
* Changed $mappings to $entities and remove logic for defining if not present.
* Change $expectedOnly to work over entities and support both a global and per-property definition.
* Fix bug where local binding of null would result in looking up local variables in global scope.
* Integrated ambiguity processing with other entity processing to make it more robust.
* Replace AdaptiveEvents.Ask with AdaptiveEvents.EndOfActions.
* Fix bug where @ would not work with anything after the primary property because first() was being applied beyond the first entity.

### CODE REVIEW

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.ApplicationInsights.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.ApplicationInsights.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.AspNet.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.AspNet.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Testing.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Testing/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Streaming.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Streaming/4.6.3)

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.ApplicationInsights.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.ApplicationInsights.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.AspNet.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.AspNet.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Testing.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Testing/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Streaming.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Streaming/4.6.3)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 1]
```diff
@@ -661,7 +661,7 @@ private async Task<bool> ProcessQueuesAsync(SequenceContext sequenceContext, Can
             }
             else
             {
-                evt = new DialogEvent() { Name = AdaptiveEvents.Ask, Bubble = false };
+                evt = new DialogEvent() { Name = AdaptiveEvents.EndOfActions, Bubble = false };
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs]: >EndOfActions [](start = 64, length = 12)

EndOfPlan?

Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs]: Considered that, but thought we had moved away from the word plan.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/PropertySchema.cs [hunk 2]
```diff
@@ -116,7 +108,13 @@ public PropertySchema(string path, JObject schema, List<PropertySchema> children
         /// Gets mappings to possible entities for property.
         /// </summary>
         /// <value>List of entity names.</value>
-        public IReadOnlyList<string> Mappings { get; }
+        public IReadOnlyList<string> Entities { get; }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/PropertySchema.cs]: do these get serialized?  I see there is a JsonIgnore attribute, in which case, all of these properties should have JsonProperty attribtes with camelcase names on them.

Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/PropertySchema.cs]: Does not get serialized.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 3]
```diff
@@ -401,7 +401,7 @@ protected virtual async Task<bool> ProcessEventAsync(SequenceContext sequenceCon
 
                         break;
 
-                    case AdaptiveEvents.EndOfActions:
```
Comment (Stevenic) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs]: Why the change? And you added "endOfPlan" so "endOfActions" is still there. What's the difference

File: libraries/Microsoft.Bot.Expressions/Memory/SimpleObjectMemory.cs [hunk 4]
```diff
@@ -36,9 +36,9 @@ public static IMemory Wrap(object obj)
         public bool TryGetValue(string path, out object value)
         {
             value = null;
-            if (memory == null)
+            if (memory == null || path.Count() < 1 || (path[0] != '[' && !char.IsLetter(path[0])))
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Expressions/Memory/SimpleObjectMemory.cs]: >Count [](start = 39, length = 5)

path.Length == 0?

Comment (tomlm) [reply libraries/Microsoft.Bot.Expressions/Memory/SimpleObjectMemory.cs]: what is this check for?

---
In reply to: [363905784](https://github.com/microsoft/botbuilder-dotnet/pull/3226#discussion_r363905784) [](ancestors = 363905784)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs (hunk 3 above)
Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs]: There are two events:
1) The sequence of actions has ended in which case we want to surface ambiguity events.  This is EndOfPlan and primariliy internal.
2) There are no more ambiguity events and we want to surface EndOfActions to possibly do prompts.

---
In reply to: [363905063](https://github.com/microsoft/botbuilder-dotnet/pull/3226#discussion_r363905063) [](ancestors = 363905063)

**Comment by tomlm** [comment]
>     public class OnEndOfActions : OnDialogEvent

Do we need a OnEndOfPlan trigger?  What is relationship between EndOfActions<=>EndOfPlan?

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TriggerConditions/OnEndOfActions.cs:13 in 77050a4. [](commit_id = 77050a4d424b8e8624fa07029b16519ee5a5532a, deletion_comment = False)

File: libraries/Microsoft.Bot.Expressions/Memory/SimpleObjectMemory.cs (hunk 4 above)
Comment (chrimc62) [reply libraries/Microsoft.Bot.Expressions/Memory/SimpleObjectMemory.cs]: Local paths cannot start with anything other than a letter.  ==0 would be fine and I can change that.

---
In reply to: [363906193](https://github.com/microsoft/botbuilder-dotnet/pull/3226#discussion_r363906193) [](ancestors = 363906193,363905784)

**Comment by tomlm** [comment]
>         public const string AssignEntity = "assignEntity";

we should describe these here...

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveEvents.cs:16 in 77050a4. [](commit_id = 77050a4d424b8e8624fa07029b16519ee5a5532a, deletion_comment = False)

**Comment by tomlm** [comment]
>     /// Triggered to ask for a any triggered prompt.

update comment

---
Refers to: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TriggerConditions/OnEndOfActions.cs:11 in b1dc7a8. [](commit_id = b1dc7a89fdb165f9af3e13d07e12b7dd9cf9d909, deletion_comment = False)

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.ApplicationInsights.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.ApplicationInsights.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.AspNet.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.AspNet.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Testing.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Testing/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Streaming.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Streaming/4.6.3)
